### PR TITLE
feat: parse `BUILD.bzel` to determine whether a commit that only changed `BUILD.bazel` is a qualified commit

### PR DESCRIPTION
**Pull request by JoeWang1127**
In this PR:
- For a commit that only changed `BUILD.bzel`, parse the `BUILD.bazel` to determine whether this is a qualified commit. The commit is a qualified commit if the two `Gapic_Inputs` objects parsed from the commit and its parent are different.

### CODE REVIEW

File: library_generation/model/config_change.py [hunk 1]
```diff
@@ -156,3 +170,30 @@ def __create_qualified_commit(
         if len(libraries) == 0:
             return None
         return QualifiedCommit(commit=commit, libraries=libraries)
+
+    @staticmethod
+    def __qualified_build_change(commit: Commit, build_file_path: str) -> bool:
```
Comment (diegomarquezp) [inline library_generation/model/config_change.py:175]: readability nit: `__is_qualified_build_change` may help to quickly understand the check in line 159.

File: library_generation/model/config_change.py [hunk 2]
```diff
@@ -156,3 +170,30 @@ def __create_qualified_commit(
...
+        """
+        Checks if the given commit containing a BUILD.bazel change is a
+        qualified commit.
```
Comment (diegomarquezp) [inline library_generation/model/config_change.py:178]: Can you add another sentence explaining that the function parses the `gapic_inputs` from both versions to tell if there is a relevant change?

File: library_generation/model/config_change.py [hunk 3]
```diff
@@ -143,11 +146,22 @@ def __create_qualified_commit(
         :return: qualified commits.
         """
         libraries = set()
+        file_change_num = len(commit.stats.files.keys())
         for file in commit.stats.files.keys():
-            if file.endswith("BUILD.bazel"):
-                continue
             versioned_proto_path = find_versioned_proto_path(file)
             if versioned_proto_path in proto_paths:
+                # determine a commit that only contains `BUILD.bazel`
```
Comment (diegomarquezp) [inline library_generation/model/config_change.py:153]: I think the `if` statement this comment refers to _will skip the commit if the only affected file in the commit is the BUILD file and doesn't contain relevant changes_. Can you please explain that in the comment?

**Review by diegomarquezp — DISMISSED** [review]
Thank you. Just a few small nits.

File: library_generation/model/config_change.py (hunk 2 above)
Comment (JoeWang1127) [reply library_generation/model/config_change.py:179]: Done.

File: library_generation/model/config_change.py (hunk 1 above)
Comment (JoeWang1127) [reply library_generation/model/config_change.py]: Done.

File: library_generation/model/config_change.py (hunk 3 above)
Comment (JoeWang1127) [reply library_generation/model/config_change.py]: Done.

File: library_generation/model/config_change.py [hunk 4]
```diff
@@ -143,11 +146,23 @@ def __create_qualified_commit(
...
-            if file.endswith("BUILD.bazel"):
-                continue
             versioned_proto_path = find_versioned_proto_path(file)
             if versioned_proto_path in proto_paths:
+                # Skip a commit if the only changed file is `BUILD.bazel`
+                # and the change doesn't contain fields used in library
+                # generation, e.g., transport, rest_numeric_enum.
+                if (
+                    file.endswith("BUILD.bazel")
+                    and file_change_num == 1
```
Comment (blakeli0) [inline library_generation/model/config_change.py:158]: What if a commit contains changes to multiple `BUILD.bazel` files? It is a common use case for multi-module library.

File: library_generation/model/config_change.py [hunk 5]
```diff
@@ -156,3 +171,31 @@ def __create_qualified_commit(
...
+        parsed from the commit and its parent are different, since there are
+        changes in fields that used in library generation.
+
+        :param commit: a GitHub commit object.
+        :param build_file_path: the path of the BUILD.bazel
+        :return: True if the commit is a qualified commit; False otherwise.
+        """
+        versioned_proto_path = find_versioned_proto_path(build_file_path)
+        build = str((commit.tree / build_file_path).data_stream.read())
+        parent_commit = commit.parents[0]
```
Comment (blakeli0) [inline library_generation/model/config_change.py:192]: I know we are getting it from `commit.parents`, but conceptually I guess this is the last/previous commit?

**Review by blakeli0 — COMMENTED** [review]
Is there going to be another PR to actually creates the PR description or it should be already taken care of by existing logics?

**Comment by JoeWang1127** [comment]
> Is there going to be another PR to actually creates the PR description or it should be already taken care of by existing logics?

The existing logic should take care of creating pr description as long as the commit is marked as a qualified commit.

File: library_generation/model/config_change.py (hunk 4 above)
Comment (JoeWang1127) [reply library_generation/model/config_change.py:158]: Good catch, I need to make some changes when a commit contains multiple `BUILD.bazel`.

File: library_generation/model/config_change.py (hunk 5 above)
Comment (JoeWang1127) [reply library_generation/model/config_change.py:199]: Maybe.

However, there might be multiple commits changed a `BUILD.bazel` between baseline and current commit.

Comment (blakeli0) [reply library_generation/model/config_change.py:199]: > there might be multiple commits changed a BUILD.bazel between baseline and current commit

Can you elaborate on this with an example?

File: library_generation/model/config_change.py [hunk 6]
```diff
@@ -143,11 +149,23 @@ def __create_qualified_commit(
...
         """
         libraries = set()
-        for file in commit.stats.files.keys():
-            if file.endswith("BUILD.bazel"):
-                continue
+        for file, changes in commit.stats.files.items():
             versioned_proto_path = find_versioned_proto_path(file)
             if versioned_proto_path in proto_paths:
+                if (
+                    file.endswith("BUILD.bazel")
```
[the remark this answers is not in the document]
Comment (JoeWang1127) [reply library_generation/model/config_change.py:156]: The file is a relative path starting from `google`, e.g., `google/cloud/aiplatform/v1/BUILD.bazel`.

File: library_generation/test/model/config_change_unit_tests.py [hunk 7]
```diff
@@ -208,6 +208,29 @@ def test_get_qualified_commits_success(self):
             qualified_commits[2].commit.hexsha,
         )
 
+    def test_get_qualified_commits_build_only_commit_returns_a_commit(self):
```
[the remark this answers is not in the document]
Comment (JoeWang1127) [reply library_generation/test/model/config_change_unit_tests.py]: I changed the test name to `test_get_qualified_commits_rest_numeric_enum_change_returns_a_qualified_commit`. WDYT?

File: library_generation/test/model/config_change_unit_tests.py [hunk 8]
```diff
@@ -228,8 +251,35 @@ def test_get_qualified_commits_build_only_commit_returns_empty_list(self):
         )
         # one commit between latest_commit and baseline_commit which only
         # changed BUILD.bazel.
+        # this commit didn't change fields used in library generation.
         self.assertTrue(len(config_change.get_qualified_commits()) == 0)
 
+    def test_get_qualified_commits_new_client_commit_returns_a_commit(self):
```
[the remark this answers is not in the document]
Comment (JoeWang1127) [reply library_generation/test/model/config_change_unit_tests.py]: I changed the test name to `test_get_qualified_commits_add_build_file_returns_a_qualified_commit`. WDYT?

File: library_generation/model/config_change.py (hunk 5 above)
Comment (JoeWang1127) [reply library_generation/model/config_change.py:199]: > Can you elaborate on this with an example?

I don't have a concrete example but now I understand your original question.

Yes, the parent commit contains the last commit touching this file.

File: library_generation/model/config_change.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (blakeli0) [reply library_generation/model/config_change.py:156]: Makes sense. A nit improvement, rename `file` to `file_path`, I thought it is a `file_name` that is only `BUILD.bazel`.

**Review by blakeli0 — APPROVED** [review]

**Review by diegomarquezp — APPROVED** [review]

**Pull request merged**